### DISCUSSION TITLE
ProgressBar animation double my processing time! Is there a workaround?

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#15525] [General]
Hi, I just found that removing the indeterminate ProgressBar animation while my app is processing had a such HUGE impact on the performance!

I am doing ML processing using torchsharp, and for the particular model I am using, the computation time is 20 sec. But, if I remove the progress bar that I show during processing... it is only 10 sec!!!

I am very surprised that ProgressBar (and animation in general?) is so time consuming... Is there a simple trick to improve the performances? For example, is it possible to skip frames in the animation? 

I would like to have a progress bar (or progress ring) to show that my application is busy, but that's a problem if the animation double the processing time!

Any idea? I do not have noticed that before, but to me it is a problem if the "busy" control of the framework slow down the application.

**Comment by rabbitism** [comment]
hmm, are you computing in UI thread?

**Reply by dbriard** [reply]
No, in another thread with Task.Factory.StartNew(...)

Finally, I think it is more complicated than that because I cannot reproduce the problem in a repro sample. I have 4 progresses running at the screen, they are not the original avalonia's FluentTheme, inside ViewBoxes, and there is probably overdraw (there are images, shadows etc... at screen)...

I will continue to investigate.

**Reply by Mrxx99** [reply]
Are you reporting progress from your thread to the progress bar? If yes it always has to dispatch to the UI thread. You can limit that to e.g. report progress not several times a second but e.g. only report new progress if the percentage increased by  1 or more percent since last reporting

**Comment by robloo** [comment]
Also note that the `ProgressRing` was abandoned due to performance issues:

 * https://github.com/AvaloniaUI/Avalonia/pull/6003#issuecomment-951458944 Discussing some of the potential causes.
 * https://github.com/AvaloniaUI/Avalonia/pull/6003
 * https://github.com/AvaloniaUI/Avalonia/pull/6792

Side note partially off-topic:
I suspect the animation system isn't quite as efficient as WPF in areas. Namely Avalonia is designed to use non-mutable structs all over the place whereas WPF could just re-use and mutate existing classes for animation. I've long suspected that "optimization" in Avalonia would be an issue with the animation system performance.

**Comment by dbriard** [comment]
Thanks guys for helping!

After investigation, I cannot really reproduce the issue... I don't know what happened, as in the first place it was clear than when removing the progress bar or progress ring, the computation time decreased by a visible margin, but not anymore today... 

May be I had to many apps running on the PC and that had an effect on the performances, especially when the animation was running... its a mystery to me! (I only have a chipset integrated GPU).

On Friday I had the impression that when the ProgressBar was on an image, it had an effect on the perfs (~13-14s vs 11-10s), but not today... 
![image](https://github.com/AvaloniaUI/Avalonia/assets/3595443/4e016496-e8dd-410b-a317-3b859d288485)
I even compared with WPF and the behavior was the same than with Avalonia on Friday: a little slower with progress bar. By the way, the Avalonia progress bar is really smoother than the default's one of WPF ;)

@robloo That what I was afraid of, but I also tested with ProgressRing (based on FluentAvalonia) and it do not look slower than progress bar...

FYI: @Mrxx99 @jmacato Avalonia 11.0.10 and I was not reporting progress, just has the indeterminate progress bar.
